File: app/src/main/java/com/androidpi/literefresh/sample/base/model/Resource.java
/*
 * Copyright 2018 yinpinjiu@gmail.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.androidpi.literefresh.sample.base.model;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

public class Resource<T> {

    /**
     * Status of a resource that is provided to the UI.
     * <p>
     * These are usually created by the Repository classes where they return
     * {@code LiveData<Resource<T>>} to pass back the latest data to the UI with its fetch status.
     */
    public enum Status {
        SUCCESS,
        ERROR,
        LOADING
    }

    @NonNull
    public final Status status;

    @Nullable
    public final String message;

    @Nullable
    public final T data;

    @Nullable
    public final Throwable throwable;

    public Resource(@NonNull Status status, @Nullable T data, @Nullable String message, @Nullable Throwable throwable) {
        this.status = status;
        this.data = data;
        this.message = message;
        this.throwable = throwable;
    }


    public static <T> Resource<T> loading() {
        return new Resource<>(Status.LOADING, null, null, null);
    }

    public static <T> Resource<T> loading(T data) {
        return new Resource<>(Status.LOADING, data, null, null);
    }

    public static <T> Resource<T> success() {
        return new Resource<>(Status.SUCCESS, null, null, null);
    }

    public static <T> Resource<T> success(T data) {
        return success(data, null);
    }

    public static <T> Resource<T> success(T data, String message) {
        return new Resource<>(Status.SUCCESS, data, message, null);
    }

    public static <T> Resource<T> error() {
        return new Resource<>(Status.ERROR, null, null, null);
    }

    public static <T> Resource<T> error(@Nullable String msg, @Nullable T data, @Nullable Throwable throwable) {
        return new Resource<>(Status.ERROR, data, msg, throwable);
    }

    public static <T> Resource<T> error(@NonNull String msg) {
        return error(msg, null, null);
    }

    public static <T> Resource<T> error(@NonNull Throwable throwable) {
        return error(null, null, throwable);
    }

    public boolean isError() {
        return status == Status.ERROR;
    }

    public boolean isSuccess() {
        return status == Status.SUCCESS;
    }

    public boolean isLoading() {
        return status == Status.LOADING;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        Resource<?> resource = (Resource<?>) o;

        if (status != resource.status) {
            return false;
        }
        if (message != null ? !message.equals(resource.message) : resource.message != null) {
            return false;
        }
        if (data != null ? !data.equals(resource.data) : resource.data != null) {
            return false;
        }
        return throwable != null ? throwable.equals(resource.throwable) : resource.throwable == null;
    }

    @Override
    public int hashCode() {
        int result = status.hashCode();
        result = 31 * result + (message != null ? message.hashCode() : 0);
        result = 31 * result + (data != null ? data.hashCode() : 0);
        result = 31 * result + (throwable != null ? throwable.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "Resource{" +
                "status=" + status +
                ", message='" + message + '\'' +
                ", data=" + data +
                ", throwable=" + throwable +
                '}';
    }
}

File: samples/src/main/java/literefresh/sample/base/model/Resource.kt
/*
 * Copyright 2018 yinpinjiu@gmail.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package literefresh.sample.base.model

class Resource<T>(val status: Status, val data: T?, val message: String?, val throwable: Throwable?) {
    /**
     * Status of a resource that is provided to the UI.
     *
     *
     * These are usually created by the Repository classes where they return
     * `LiveData<Resource<T>>` to pass back the latest data to the UI with its fetch status.
     */
    enum class Status {
        SUCCESS, ERROR, LOADING
    }

    val isError: Boolean
        get() = status == Status.ERROR

    val isSuccess: Boolean
        get() = status == Status.SUCCESS

    val isLoading: Boolean
        get() = status == Status.LOADING

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val resource = o as Resource<*>
        if (status != resource.status) {
            return false
        }
        if (if (message != null) message != resource.message else resource.message != null) {
            return false
        }
        if (if (data != null) data != resource.data else resource.data != null) {
            return false
        }
        return if (throwable != null) throwable == resource.throwable else resource.throwable == null
    }

    override fun hashCode(): Int {
        var result = status.hashCode()
        result = 31 * result + (message?.hashCode() ?: 0)
        result = 31 * result + (data?.hashCode() ?: 0)
        result = 31 * result + (throwable?.hashCode() ?: 0)
        return result
    }

    override fun toString(): String {
        return "Resource{" +
                "status=" + status +
                ", message='" + message + '\'' +
                ", data=" + data +
                ", throwable=" + throwable +
                '}'
    }

    companion object {
        fun <T> loading(): Resource<T> {
            return Resource(Status.LOADING, null, null, null)
        }

        fun <T> loading(data: T): Resource<T> {
            return Resource(Status.LOADING, data, null, null)
        }

        fun <T> success(): Resource<T?> {
            return Resource(Status.SUCCESS, null, null, null)
        }

        fun <T> success(data: T): Resource<T> {
            return success(data, null)
        }

        fun <T> success(data: T, message: String?): Resource<T> {
            return Resource(Status.SUCCESS, data, message, null)
        }

        fun <T> error(): Resource<T> {
            return error(null, null, null)
        }

        fun <T> error(msg: String?, data: T?, throwable: Throwable?): Resource<T> {
            return Resource(Status.ERROR, data, msg, throwable)
        }

        fun <T> error(msg: String): Resource<T> {
            return error(msg, null, null)
        }

        fun <T> error(throwable: Throwable): Resource<T> {
            return error(null, null, throwable)
        }
    }

}